### PR TITLE
Require exhaustive matches

### PR DESCRIPTION
**Pull request by timriley**
This changes makes it required for a match block to provide handlers for all the cases in the matcher. For example, when using `Dry::Matcher::EitherMatcher`, both `success` and `failure` blocks must be provided. A failure to handle all the cases will result in a `NonExhaustiveMatchError`.

This change should make matcher usage more robust, since it less likely that certain conditions are forgotten/ignored.

Since this is a behavioural change, I'd be happy to hear thoughts about this.

In our usage of dry-matcher at Icelab, this would continue to work fine for us, since in 99% of our cases we already provide success/failure handlers, and in the 1%, we should probably just go and add the missing one :)

### CODE REVIEW

**Comment by solnic** [comment]
Yes this is a desired behavior. ie in Scala when you use Option and don't provide matching block for `None` your code won't compile.

File: lib/dry/matcher/evaluator.rb [hunk 1]
```diff
@@ -20,11 +29,18 @@ def respond_to_missing?(name, include_private = false)
...
         return super unless @cases.key?(name)
 
+        @handled.add name
         handle_case @cases[name], *args, &block
       end
 
       private
 
+      def ensure_exhaustive_match
+        if (not_handled = @cases.keys - @handled.to_a).any?
```
Comment (flash-gordon) [inline lib/dry/matcher/evaluator.rb]: You can also treat all cases as not handled initially and then remove them one by one. This won't give much but you'll be able to simplify this expression to `if @not_handled.any?`

File: spec/integration/matcher_spec.rb [hunk 2]
```diff
@@ -57,6 +74,8 @@ def call_match(input)
 
       def call_match(input)
         matcher.(input) do |m|
+          m.success { }
```
Comment (flash-gordon) [inline spec/integration/matcher_spec.rb]: I think we shouldn't force the user to pass an empty block, wdyt?

Comment (timriley) [inline spec/integration/matcher_spec.rb]: Turns out we don't force this already! I'll clear up this test :)

File: lib/dry/matcher/evaluator.rb (hunk 1 above)
Comment (timriley) [inline lib/dry/matcher/evaluator.rb]: Oh yes, great idea! Pushing this up now!

**Comment by ushis** [comment]
It would be amazing provide some sort of wildcard method, that matches all unhandled cases. Think about a matcher with 4 cases ```:success, :invalid, :not_found, :unauthorized``` and a caller caring only about 2 of them and likes to process the rest in a generic way:

```ruby
matcher.(result) do |m|
  m.success { |v| "Yay! You did something: #{v}" }
  m.invalid { |v| "Hmm... You did something, but wrong: #{v.errors}" }
  m.else { |v| "No! Something went really wrong..." }
end
```

@timriley  what do you think?

**Pull request merged**